### PR TITLE
Kimi K2.6 Dflash in prefill runner

### LINKED ISSUE #50569: DFlash integration into prefill runner
once DFlash works on 1GLX we an integrate it into runner so it can be used in a pipeline

### PR DESCRIPTION
**Pull request by nbabinTT**
### PR Category
Feature

### Summary
DFlash is Kimi-K2.6's speculative-decode drafter. The device drafter already landed and was validated on its own, but it had no way into a real run — the integration test wired up the transformer by hand, there was no on/off switch, and nothing built the drafter's KV cache during an actual prefill. This PR closes that gap. With `PREFILL_DFLASH=1`, an ordinary Kimi prefill launched through the production runner now builds the drafter's context-KV cache as it goes, pipelined across galaxies. The flag defaults off and rides the same config chain as `kv_only_last_layer`; the checkpoint comes from `$DFLASH_HF_MODEL`. Nothing changes for any other model when it's off.

The core of the change is spreading the drafter's FC context projection over the pipeline. Instead of collecting all six target-layer hiddens and running one matmul at the end, each rank owns only the target layers in its slice and streams `fc_slice_i @ h_i` as those layers are tapped. The partials sum by linearity — `Σ reduce_scatter(A_r) == reduce_scatter(Σ A_r)` — so no rank needs the full set, and the last rank finalizes into the caches (hidden_norm → per-draft-layer K/V). Between galaxies the partial travels on the D2D socket we already use for the hidden state, packed beside it as a 2·H-wide activation and split apart on the receiver; the runner only widens the handoff when DFlash is on.

This also drops the old `concat` FC mode, which stored the raw taps for one grouped matmul at write time. That path can't cross ranks — it would forward 6× the data — so sliced is now the only mode. It measured higher PCC anyway.

Closes #50569.

### What changed
- **Drafter** (`tt_dflash_drafter.py`) — sliced-only FC; per-rank `owned_target_layer_ids` + `build_kv_tail`; `import_partial` / `export_partial` / `_finalize_sharded_partial` for the pipeline handoff. The `concat` mode and its stored-tap bookkeeping are gone.
- **Runtime** (`tt_prefill_runtime.py`) — builds the drafter under the flag, taps each owned target layer through the existing `on_layer_hidden` hook, packs/unpacks the 2·H activation, and on the last rank allocates and fills the caller-owned KV caches.
- **Runner** (`prefill_runner.py`, `adapter.py`, `adapters/mla.py`) — the `PREFILL_DFLASH` env and `dflash_enabled` param, plus the D2D activation width bumped to 2·H only when enabled. The runner stays model-agnostic.
- **Config + weights** (`dflash_drafter_config.py`, new `utils.py`) — `from_hf_config` / `from_pretrained`, the rope-config helper moved into `utils.py`, and a selective safetensors loader that reads only the tensors a rank needs (non-tail ranks load just `fc.weight`).
- **Tests + CI** — the standalone and integration tests drop the `fc_mode` axis, and the `kimi_dflash` blaze leg follows.

### Testing (Blackhole)
- `test_dflash_pcc` on an 8×4 Galaxy (`kimi_dflash` CI leg) — device drafter context-KV vs the real HF `DFlashDraftModel`, sliced FC, pretrained weights.
- The 61-layer integration test taps the six target layers through the runtime hook and PCCs the finalized caches against the HF reference (0.999 threshold).
- End to end through the runner on hardware, single-rank and across two galaxies over the packed D2D handoff.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nbabin/dflash_multi_host_rebase)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nbabin/dflash_multi_host_rebase)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nbabin/dflash_multi_host_rebase)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nbabin/dflash_multi_host_rebase)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nbabin/dflash_multi_host_rebase)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nbabin/dflash_multi_host_rebase)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nbabin/dflash_multi_host_rebase)

### CODE REVIEW

**Review by mtairum — APPROVED** [review]

File: models/demos/common/prefill/adapter.py [hunk 1]
```diff
@@ -66,6 +66,7 @@ class PrefillRunParams:
     num_links: int
     gate_mode_name: str  # GateComputeMode enum name
     kv_only_last_layer: bool
+    dflash_enabled: bool  # build the DFlash drafter context-KV cache
     weight_cache_path: Optional[Path]
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply models/demos/common/prefill/adapter.py]: Yes, let's have this false by default

File: models/demos/common/prefill/runners/prefill_runner.py [hunk 2]
```diff
@@ -179,6 +179,11 @@ def sd(key, val):
 # When on (default), the last transformer layer runs kv-only: it fills the KV cache for migration and
 # skips its Q/SDPA/wo, FFN/MoE, final norm, and LM head. In a pipeline only the last rank applies it.
 KV_ONLY_LAST_LAYER = os.environ.get("PREFILL_KV_ONLY_LAST_LAYER", "1") == "1"
+# Build the DFlash drafter context-KV cache during this prefill. There is no dedicated on/off env: dflash
+# runs when the selected model declares the capability (ADAPTER.supports_dflash) AND a drafter checkpoint is
+# provided (DFLASH_HF_MODEL, resolved by the runtime). This keeps a non-dflash model from ever building a
+# Kimi drafter, and makes the checkpoint's presence the switch (DFLASH_HF_MODEL= empty forces it off).
+DFLASH_ENABLED = ADAPTER.supports_dflash and bool(os.environ.get("DFLASH_HF_MODEL"))
```
Comment (pavlepopovic) [inline models/demos/common/prefill/runners/prefill_runner.py]: per offline discussion, let's add assert in runner if DFLASH is enabled, that the USE_TRACE must be false

File: models/demos/common/prefill/runners/prefill_runner.py [hunk 3]
```diff
@@ -817,6 +822,11 @@ def _print_config() -> None:
         ("PREFILL_NUM_LAYERS", str(NUM_LAYERS)),
         ("PREFILL_PP_LAYER_COUNTS", os.environ.get("PREFILL_PP_LAYER_COUNTS", "<even split>")),
         ("PREFILL_KV_ONLY_LAST_LAYER", str(KV_ONLY_LAST_LAYER)),
+        (
+            "DFLASH_ENABLED",
+            f"{DFLASH_ENABLED} (adapter.supports_dflash={ADAPTER.supports_dflash}, "
```
Comment (pavlepopovic) [inline models/demos/common/prefill/runners/prefill_runner.py:661]: shouldn't we also check if DFLASH env var is set ?

File: models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py [hunk 4]
```diff
@@ -499,19 +602,37 @@ def on_layer_complete(layer_idx: int) -> None:
...
             cache_user_id=slot_id,
             index_kv_cache=kv_caches.index,
         )
-        ttnn.deallocate(input_tensor)
+        ttnn.deallocate(model_input)
+
+        if self.config.dflash_enabled:
+            if self.config.is_last_rank:
+                # Finalize the drafter context-KV into the runtime-owned caches (read back for PCC / migration).
+                self.drafter.write_kv_cache(self._dflash_k_cache, self._dflash_v_cache)
```
Comment (pavlepopovic) [inline models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py]: can we rename this to something more descriptive? Like forward or something ?

**Review by pavlepopovic — COMMENTED** [review]
Address AI comments aswell pls they seem good :)

File: models/demos/common/prefill/adapter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (philei-tt) [reply models/demos/common/prefill/adapter.py]: Also, should this be in common in the the place? Not all models support dflash, some support other spec decodes, right?

File: models/demos/common/prefill/runners/prefill_runner.py (hunk 3 above)
Comment (nbabinTT) [reply models/demos/common/prefill/runners/prefill_runner.py:661]: done

**Review by pavlepopovic — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py (hunk 4 above)
Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py]: renamed to 'forward'

File: models/demos/common/prefill/adapter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbabinTT) [reply models/demos/common/prefill/adapter.py]: I think we should have some attribute in model adapter which tells us if a model supports spec decode. I suggest generalizing this in a separate PR as multiple spec decode implementations will land in the future, so we will have to find some elegant way to implement this.

File: tests/pipeline_reorg/blaze_models_prefill_tests.yaml [hunk 5]
```diff
@@ -512,7 +512,7 @@
 
     mpirun --bind-to none --pernode --tag-output bash -lc '
       export OMP_NUM_THREADS=$(nproc)
-      python3 -m pytest models/demos/deepseek_v3_d_p/tests/dflash_prefill/test_dflash.py::test_dflash_pcc -k "mesh-8x4 and sliced and pretrained" -vvv --tb=short --timeout=600
+      python3 -m pytest models/demos/deepseek_v3_d_p/tests/dflash_prefill/test_dflash.py::test_dflash_pcc -k "mesh-8x4 and pretrained" -vvv --tb=short --timeout=600
```
[the remark this answers is not in the document]
Comment (nbabinTT) [reply tests/pipeline_reorg/blaze_models_prefill_tests.yaml:550]: this will be addressed in https://github.com/tenstorrent/tt-metal/pull/52413

File: models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py [hunk 6]
```diff
@@ -208,16 +221,106 @@ def _build_model(self, state_dict: dict) -> None:
...
+            self.drafter.tap(h, global_idx)
+
+        self._on_layer_hidden = on_layer_hidden
+
+        # Only the last rank finalizes the drafter KV → allocate the caller-owned context caches it fills.
+        if self.config.is_last_rank:
+            self._dflash_k_cache, self._dflash_v_cache = allocate_dflash_kv_cache(
+                self.mesh_device,
+                dcfg,
+                dflash_seq,  # chunk_size — MUST match the drafter's cache_seq/rope (see note above)
```
[the remark this answers is not in the document]
Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tt/tt_prefill_runtime.py:290]: addressed in https://github.com/tenstorrent/tt-metal/pull/52413

File: models/demos/deepseek_v3_d_p/tests/dflash_prefill/test_dflash_prefill_integration.py [hunk 7]
```diff
@@ -276,14 +275,14 @@ def _to_host(t):  # -> host [1, seq, H]; TP concatenated on hidden → full H
         )
         return host.reshape(1, isl_total, H).float()
 
-    target_hiddens = [_to_host(drafter._taps[j]) for j in range(len(dcfg.target_layer_ids))]
+    target_hiddens = [_to_host(tapped_hiddens[tid]) for tid in dcfg.target_layer_ids]
```
[the remark this answers is not in the document]
Comment (nbabinTT) [reply models/demos/deepseek_v3_d_p/tests/dflash_prefill/test_dflash_prefill_integration.py:278]: done

File: models/demos/common/prefill/runners/prefill_runner.py (hunk 2 above)
Comment (nbabinTT) [reply models/demos/common/prefill/runners/prefill_runner.py]: done

File: models/demos/common/prefill/runners/prefill_runner.py [hunk 8]
```diff
@@ -943,6 +947,9 @@ def main() -> None:
         # headless: its last layer runs KV-only and no norm/LM-head is built. Only the last rank does
         # this (single-rank inherits it); PREFILL_KV_ONLY_LAST_LAYER can force it off.
         kv_only_last_layer=is_last_rank and KV_ONLY_LAST_LAYER,
+        # NOT gated on is_last_rank: every rank builds its owned fc slices; the runtime derives the
+        # last-rank KV tail from is_last_rank.
+        dflash_enabled=DFLASH_ENABLED,
```
[the remark this answers is not in the document]
Comment (nbabinTT) [reply models/demos/common/prefill/runners/prefill_runner.py:794]: done
